File: libs/portal/src/cube/cube-service.ts
import {HTTPService} from "@portal/communication";
import { CubeDescriptor } from "./cube-metadata";


export class CubeService extends HTTPService {
   constructor(options?: any) {
       super();
    this.baseUrl = options?.baseUrl ?? "http://localhost:8000/api/cube/";
  }

  /** POST /create */
  createCube(cube: CubeDescriptor): Promise<CubeDescriptor> {
    return this.post<CubeDescriptor, CubeDescriptor>("create", cube);
  }

  /** POST /update */
  updateCube(cube: CubeDescriptor): Promise<CubeDescriptor> {
    return this.post<CubeDescriptor, CubeDescriptor>("update", cube);
  }

  /** POST /list */
  listCubes(): Promise<CubeDescriptor[]> {
    return this.get<CubeDescriptor[]>("list");
  }

  /** POST /deploy */
  async deployCube(cube: CubeDescriptor): Promise<void> {
    await this.post<CubeDescriptor, unknown>("deploy", cube);
  }
}

File: server/packages/cube/src/cube/interface/cube_service.py
from abc import abstractmethod
from typing import List
from aspyx_service import service, Service, rest, get, post, Body

from .cube_descriptors import CubeDescriptor


@service(name="cube-service", description="metadata stuff")
@rest("/api/cube/")
class CubeService(Service):
    """
    The portal service exposes cube data.
    """

    @abstractmethod
    @post("create")
    def create_cube(self, cube: Body(CubeDescriptor)) -> CubeDescriptor:
        pass

    @abstractmethod
    @post("update")
    def update_cube(self, cube: Body(CubeDescriptor)) -> CubeDescriptor:
        pass

    @abstractmethod
    @get("list")
    def list_cubes(self) -> List[CubeDescriptor]:
        pass

    @abstractmethod
    @post("deploy")
    def deploy_cube(self, cube: Body(CubeDescriptor)):
        pass